Source/Core/VideoBackends/OGL/TextureCache.cpp

```diff
@@ -588,4 +633,159 @@ void TextureCache::ConvertTexture(TCacheEntryBase* _entry, TCacheEntryBase* _unc
+                                      u32 aligned_width, u32 aligned_height, u32 row_stride,
+                                      const u8* palette, TlutFormat palette_format)
+{
+  auto key = std::make_pair(static_cast<u32>(format), static_cast<u32>(palette_format));
+  auto iter = s_texture_decoding_program_info.find(key);
+  if (iter == s_texture_decoding_program_info.end())
+    return;
+
+#ifdef TIME_TEXTURE_DECODING
+  GPUTimer timer;
+#endif
+
+  // Copy to GPU-visible buffer, aligned to the data type.
+  auto info = iter->second;
+  u32 bytes_per_buffer_elem =
+      TextureConversionShader::GetBytesPerBufferElement(info.base_info->buffer_format);
+
+  // Only copy palette if it is required.
+  bool has_palette = info.base_info->palette_size > 0;
+  u32 total_upload_size = static_cast<u32>(data_size);
+  u32 palette_offset = total_upload_size;
+  if (has_palette)
+  {
+    // Align to u16.
+    if ((total_upload_size % sizeof(u16)) != 0)
+    {
+      total_upload_size++;
+      palette_offset++;
+    }
+
+    total_upload_size += info.base_info->palette_size;
+  }
+
+  // Allocate space in stream buffer, and copy texture + palette across.
+  auto buffer = s_palette_stream_buffer->Map(total_upload_size, bytes_per_buffer_elem);
+  memcpy(buffer.first, data, data_size);
+  if (has_palette)
+    memcpy(buffer.first + palette_offset, palette, info.base_info->palette_size);
+  s_palette_stream_buffer->Unmap(total_upload_size);
+
+  info.program.Bind();
+
+  // Calculate stride in buffer elements
+  u32 row_stride_in_elements = row_stride / bytes_per_buffer_elem;
+  u32 offset_in_elements = buffer.second / bytes_per_buffer_elem;
+  u32 palette_offset_in_elements = (buffer.second + palette_offset) / sizeof(u16);
+  if (info.uniform_dst_size >= 0)
+    glUniform2ui(info.uniform_dst_size, width, height);
+  if (info.uniform_src_size >= 0)
+    glUniform2ui(info.uniform_src_size, aligned_width, aligned_height);
+  if (info.uniform_src_offset >= 0)
+    glUniform1ui(info.uniform_src_offset, offset_in_elements);
+  if (info.uniform_src_row_stride >= 0)
+    glUniform1ui(info.uniform_src_row_stride, row_stride_in_elements);
+  if (info.uniform_palette_offset >= 0)
+    glUniform1ui(info.uniform_palette_offset, palette_offset_in_elements);
+
+  glActiveTexture(GL_TEXTURE9);
+  glBindTexture(GL_TEXTURE_BUFFER, s_texture_decoding_buffer_views[info.base_info->buffer_format]);
+
+  if (has_palette)
+  {
+    // Use an R16UI view for the palette.
+    glActiveTexture(GL_TEXTURE10);
+    glBindTexture(GL_TEXTURE_BUFFER, s_palette_resolv_texture);
+  }
+
+  auto dispatch_groups = TextureConversionShader::GetDispatchCount(info.base_info, width, height);
+  glBindImageTexture(0, static_cast<TCacheEntry*>(entry)->texture, dst_level, GL_TRUE, 0,
+                     GL_WRITE_ONLY, GL_RGBA8);
+  glDispatchCompute(dispatch_groups.first, dispatch_groups.second, 1);
+  glMemoryBarrier(GL_TEXTURE_UPDATE_BARRIER_BIT);
```

Comment:
This barrier might get very expensive, especially for the last mipmap levels. What do you think about moving this after decoding all levels?